typed-protocols/src/Network/TypedProtocol/Core.hs

```diff
@@ -0,0 +1,355 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE DataKinds #-}
+
+
+-- | This module defines the core of the typed protocol framework.
+--
+
+module Network.TypedProtocol.Core (
+
+  -- * Introduction
+  -- $intro
+
+
+  -- * Defining protocols
+  -- $defining
+  Protocol(..),
+
+  -- * Engaging in protocols
+  -- $using
+  PeerKind(..),
+  PeerHasAgency(..),
+  WeHaveAgency,
+  TheyHaveAgency,
+  Peer(..),
+
+  -- * Protocol proofs and tests
+  -- $tests
+  ) where
+
+-- $intro
+-- A typed protocol between two peers is defined via a state machine: a
+-- collection of protocol states and protocol messages which are transitions
+-- between those states.
+--
+-- Start from the idea that a protocol is some language of messages sent
+-- between two peers. To specify a protocol is to describe what possible
+-- sequences of messages are valid. One simple but still relatively expressive
+-- way to do this is via a state machine: starting from some initial state,
+-- all possible paths through the state machine gives the set of valid protocol
+-- traces. This then dictates what a peer participating in a protocol may
+-- produce and what it must accept.
+--
+-- In this style we have a fixed number of states and in each state there is
+-- some number of valid messages that move us on to the next state. This can be
+-- illustrated as a graph, which can be a helpful form of documentation.
+--
+-- We further constrain this idea by saying that the two peers will use the
+-- same state machine and change states in lock-step by sending\/receiving
+-- messages. In this approach, for each protocol state, the description
+-- dictates which peer has the agency to choose to send a message, while
+-- correspondingly the other must be prepared to receive the message.
+--
+-- The views of the two peers are dual. In each state one peer can send any
+-- message that is valid for the current protocol state while the other
+-- must be prepared to receive any valid message for current protocol state.
+--
+-- We can also have terminal protocol states in which neither peer has agency.
+--
+-- So part of the protocol description is to label each protocol state with
+-- the peer that has the agency in that state, or none for terminal states.
+-- We use the labels \"client\" and \"server\" for the two peers, but they are
+-- in fact symmetric.
```

Comment:
Hmm, but it's not quite that is it because when we write a `Peer` we're always local, but we have to write peers for the two sides. So we need absolute labels not relative ones.

The language here is tricky. The best I came up with was the absolute labels client/server `ClientHasAgency`/`ServerHasAgency` and then for the relative cases we have `WeHaveAgency`/`TheyHaveAgency`.

Reply:
> Hmm, but it's not quite that is it because when we write a Peer we're always local, but we have to write peers for the two sides. So we need absolute labels not relative ones.

When you write the peer itself you have a particular partition GADT with whatever name you want. If it really is a client/server protocol, you'll probably have `Client :: state -> Type`  and `Server :: state -> Type`, so you'll get

```
server :: Peer Server Client End Tr 'Idle f t
client :: Peer Client Server End Tr 'Idle f t
```

but when dealing _generally_ with protocol applications, we can leave it as local/remote.